credentials/apps/badges/utils.py

```diff
@@ -47,3 +51,28 @@ def traverse(current, keys):
             return None
 
     return traverse(current, keys)
+
+
+def get_user_data(data) -> UserData:
+    """
+    Extracts UserData object from any dataclass that contains UserData as a field.
+    Arguments:
+        data: Input dataclass object that contains UserData.
+    Returns:
+        UserData: UserData object contained within the input dataclass.
+    """
+    # Check if the data is an instance of UserData directly
+    if isinstance(data, UserData):
+        return data
+    # Inspect attributes of the dataclass
+    for attr_name, attr_value in inspect.getmembers(data):
```

Comment:
Here we can, possibly, reduce complexity with [attrs instance filtering](https://www.attrs.org/en/stable/api.html#attrs.has)

Reply:
done